ring/util.go

```diff
@@ -102,27 +102,37 @@ func WaitRingStability(ctx context.Context, r *Ring, op Operation, minStability,
 
 	// Get the initial ring state.
 	ringLastState, _ := r.GetAllHealthy(op) // nolint:errcheck
-	ringLastStateTs := time.Now()
 
-	const pollingFrequency = time.Second
-	pollingTicker := time.NewTicker(pollingFrequency)
-	defer pollingTicker.Stop()
+	// This timer will be our state
+	stabilityTimer := time.NewTimer(minStability)
+
+	stateWatch, err := r.statesWatcher.Register()
+	if err != nil {
+		return err
+	}
+	defer r.statesWatcher.UnRegister()
 
 	for {
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
-		case <-pollingTicker.C:
+		case <-stabilityTimer.C:
 			// We ignore the error because in case of error it will return an empty
 			// replication set which we use to compare with the previous state.
 			currRingState, _ := r.GetAllHealthy(op) // nolint:errcheck
 
 			if HasReplicationSetChanged(ringLastState, currRingState) {
 				ringLastState = currRingState
-				ringLastStateTs = time.Now()
-			} else if time.Since(ringLastStateTs) >= minStability {
+				stabilityTimer.Reset(minStability)
+			} else {
 				return nil
 			}
+		case <-stateWatch:
+			ringLastState, _ = r.GetAllHealthy(op) // nolint:errcheck
+			if !stabilityTimer.Stop() {
+				<-stabilityTimer.C
+			}
+			stabilityTimer.Reset(minStability)
```

Comment:
Here we reset the timer but the fact that we received a notification from the watch doesn't mean the ring has changed so that `HasReplicationSetChanged()` return true (for example, that function doesn't compare heartbeat `Timestamp` which change continuously but the watcher will notify whenever an heartbeat timestamp has changed too).